src/storage/redis.rs

```diff
@@ -0,0 +1,400 @@
+        RedisStorageBuilder::new(pool)
+    }
+}
+
+struct ForwardedMessage {
+    pub id: String,
+    pub destination: u32,
+}
+
+impl ForwardedMessage {
+    pub fn from_str_pair<S: Into<String>>(pair: S) -> Result<Self> {
+        let str_pair = pair.into();
+        let parts: Vec<&str> = str_pair.rsplitn(2, '.').collect();
+        let dest_str = parts[0].to_string();
+        let dest = dest_str.parse::<u32>()?;
+
+        Ok(Self {
+            id: parts[1].to_string(),
+            destination: dest,
+        })
+    }
+
+    pub fn from_bytes_pair(bytes: &Vec<u8>) -> Result<Self> {
+        let pair = from_utf8(bytes)?;
+        Self::from_str_pair(pair)
+    }
+
+    pub fn to_str_pair(&self) -> String {
+        format!("{}.{}", self.id, self.destination)
+    }
+}
+
+impl ToRedisArgs for ForwardedMessage {
+    fn write_redis_args<W>(&self, out: &mut W)
+    where
+        W: ?Sized + RedisWrite,
+    {
+        let pair = self.to_str_pair();
+        out.write_arg(&pair.as_bytes());
+    }
+}
+
+impl FromRedisValue for ForwardedMessage {
+    fn from_redis_value(v: &Value) -> RedisResult<Self> {
+        if let Value::Data(data) = v {
+            let ret = ForwardedMessage::from_bytes_pair(data);
+            match ret {
+                Ok(bytes) => Ok(bytes),
+                Err(err) => Err(RedisError::from((
+                    ErrorKind::ResponseError,
+                    "cannot decode a forwarded message from from pair {}",
+                    err.to_string(),
+                ))),
+            }
+        } else {
+            Err(RedisError::from((
+                ErrorKind::TypeError,
+                "expected a data type from redis",
+            )))
+        }
+    }
+}
+
+#[async_trait]
+impl Storage for RedisStorage {
+    async fn get(&self, id: &str) -> Result<Option<Message>> {
+        let mut conn = self.get_connection().await?;
+        let key = self.prefixed(Queue::Backlog(id));
+        let ret: Option<Vec<u8>> = conn.get(key).await?;
+
+        match ret {
+            Some(val) => {
+                let msg = Message::from_json(&val)?;
+                Ok(Some(msg))
+            }
+            None => Ok(None),
+        }
+    }
+
+    async fn run(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let queue = self.prefixed(Queue::Run(&msg.command));
+
+        conn.rpush(&queue, msg).await?;
+
+        // only keep `max_commands` in this queue
+        let len: isize = conn.llen(&queue).await?;
+        if len > self.max_commands {
+            conn.ltrim(&queue, self.max_commands, -1).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn forward(&self, msg: &Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+
+        // add to backlog
+        let key = self.prefixed(Queue::Backlog(&msg.id));
+        conn.set_ex(&key, msg, self.ttl as usize).await?;
```

Comment:
This need to use the ttl from the message itself. `msg.expiration` which is expiration in seconds. May be make that with an upper limit of 1 hour or something that is very big but not infinite